distributed/worker.py

```diff
@@ -460,33 +476,56 @@ def __init__(
         if validate is None:
             validate = dask.config.get("distributed.scheduler.validate")
         self.validate = validate
-
-        self._transitions = {
-            # Basic state transitions
-            ("new", "waiting"): self.transition_new_waiting,
-            ("new", "fetch"): self.transition_new_fetch,
-            ("waiting", "ready"): self.transition_waiting_ready,
-            ("fetch", "flight"): self.transition_fetch_flight,
-            ("ready", "executing"): self.transition_ready_executing,
-            ("executing", "memory"): self.transition_executing_done,
-            ("flight", "memory"): self.transition_flight_memory,
-            ("flight", "fetch"): self.transition_flight_fetch,
-            # Shouldn't be a valid transition but happens nonetheless
-            ("ready", "memory"): self.transition_ready_memory,
-            # Scheduler intercession (re-assignment)
-            ("fetch", "waiting"): self.transition_fetch_waiting,
-            ("flight", "waiting"): self.transition_flight_waiting,
-            # Errors, long-running, constrained
-            ("waiting", "error"): self.transition_waiting_done,
+        self._transitions_table = {
+            ("cancelled", "resumed"): self.transition_cancelled_resumed,
+            ("cancelled", "fetch"): self.transition_cancelled_fetch,
+            ("cancelled", "released"): self.transition_cancelled_released,
+            ("cancelled", "waiting"): self.transition_cancelled_waiting,
+            ("cancelled", "forgotten"): self.transition_cancelled_forgotten,
+            ("cancelled", "memory"): self.transition_cancelled_memory,
+            ("cancelled", "error"): self.transition_generic_error,
```

Comment:
There is an ambiguity in states I didn't resolve, yet. There is the `rescheduled` state (rather stimulus, actually) which triggers a `{"op": "reschedule", "key": key}` to the scheduler and releases a key. This happens if the user raises a `Reschedule` exception. I haven't touched this system, yet, since it feels fragile and I'm not sure how well tested it is. 

I introduced another state/stimulus called `resumed` which acts as a virtual state for transitions between an executable to a fetchable task or the other way round. `resumed` and `rescheduled` *might* be the same thing but I'm not entirely sure, yet. This wasn't a priority of mine but I could take a look if the merger of these states makes things simpler.

Anyhow, if `resumed` is ambiguous, I can also rename it. As I said, this is a state which typically should only follow a `cancelled`, i.e. after a `flight` or `executing`. In practice, this only happens if a task is currently in `flight` but the worker is asked to compute it. That likely ends in a connection failure during fetch and will then schedule the task as ready. If the fetch finishes successfully after all, we will not compute the task since we already have the task result in memory. I experimented with "on_success" and "on_failure" hooks but that made things somehow more complex.
For executing tasks this could also happen in theory but not in practice since `executing` should either end in `memory` or `error` and will never be rescheduled

See
https://github.com/fjetter/distributed/blob/63d98479b7a1d0d9fc7fc5e8c4701db8887e3384/distributed/worker.py#L1907-L1913 
and 
https://github.com/fjetter/distributed/blob/63d98479b7a1d0d9fc7fc5e8c4701db8887e3384/distributed/worker.py#L3144-L3145